### PR TITLE
torch.library.register_fake now accepts more types

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* #124089
* #124071
* __->__ #124066
* #124065
* #124064
* #123937

We allow it to accept:
- a string with the op name
- an opoverload
- a new-style custom op

If any of these are referring to a new-style custom op (created with the
custom_op decorator), then we dispatch to CustomOpDef.register_fake.
Otherwise, we do what we previously did.

Test Plan:
- new tests

### CODE REVIEW

File: torch/_library/custom_ops.py [hunk 1]
```diff
@@ -441,6 +442,7 @@ def __call__(self, *args, **kwargs):
 
 
 OPDEF_TO_LIB: Dict[str, "library.Library"] = {}
+OPDEFS: Dict[str, CustomOpDef] = {}
```
Comment (albanD) [inline torch/_library/custom_ops.py:445]: Should this be a weak value dict?

File: torch/_library/custom_ops.py [hunk 2]
```diff
@@ -468,3 +470,11 @@ def check(arg):
         yield from check(arg)
     for kwarg in kwargs.values():
         yield from check(kwarg)
+
+
+def _maybe_get_opdef(qualname):
```
Comment (albanD) [inline torch/_library/custom_ops.py:475]: qualname sounds like it should be a str object which is not true given the below...

Comment (albanD) [reply torch/_library/custom_ops.py:475]: Also it's used only in one place, should we just move it to that other file?

Comment (zou3519) [reply torch/_library/custom_ops.py:475]: > qualname sounds like it should be a str object which is not true given the below...

Ya

> Also it's used only in one place, should we just move it to that other file?

I'm trying to not pollute library.py too much

File: torch/_library/custom_ops.py (hunk 1 above)
Comment (zou3519) [reply torch/_library/custom_ops.py]: yup

**Review by albanD — APPROVED** [review]
Thanks for the update!

**Pull request closed without merging**